src/storage-backend.js

```diff
@@ -0,0 +1,655 @@
+    this.sqs = this.config.sqs;
+
+    // Maximum number of redirects to follow
+    this.redirectLimit = config.redirectLimit || 30;
+
+    // Should we allow non-SSL links in the redirect chain
+    this.ensureSSL = typeof config.ensureSSL === 'boolean' ? config.ensureSSL : true;
+
+    return this;
+  }
+
+  /**
+   * Because constructors are not able to be asyncronous, we have an async init
+   * method which will be used to do all async setup.  This should be called
+   * regardless of whether this instance will listen for incoming file-put
+   * requests.
+   */
+  async init() {
+    let data = await this.sqs.createQueue({
+      QueueName: this.copyRequestQueueName,
+    }).promise();
+
+    // The SQSConsumer rebinds the handleMessage function to it's own scope, so
+    // I want to make sure that the function has access to this object's
+    // properties.
+    let that = this;
+
+    this.copyRequestQueue = data.data.QueueUrl;
+    this.consumer = SQSConsumer.create({
+      queueUrl: this.copyRequestQueue,
+      batchSize: 10,
+      handleMessage: (message, done) => {
+        if (message.Body) {
+          let body;
+          try {
+            body = JSON.parse(message.Body);
+          } catch (err) {
+            done(err);
+          }
+          debug(`${that.id} Received put request ${body.url}`);
+          that.put.call(that, body.url).then(() => {
+            debug(`${that.id} Completed put request ${body.url}`);
+            done();
+          }, err => {
+            debug(`${that.id} Error calling put in handler ${err.stack || err}`); 
+            done(err);
+          });
+        } else {
+          debug('Received an empty message, huh?');
+          done();
+        }
+      },
+      sqs: this.sqs,
+    });
+  }
+  /**
+   * Follow redirects in a secure way, unless configured not to.  This function
+   * will ensure that all redirects in a redirect chain are pointing to HTTPS
+   * urls and not HTTP urls.  This is used to ensure that everything in the Cloud
+   * Mirror storage was obtained through with an HTTP chain of custody.  If the 
+   * config parameter 'allowInsecureRedirect' is a truthy value, HTTP urls and 
+   * redirections from HTTPS to HTTP resources will be allowed.
+   *
+   * TODO:
+   * - Ensure that the certificate validation being done here is using
+   *   certificates that we're comfortable with.
+   * - Decide whether the allowedPatterns should factor in here.  Arguably, if we
+   *   start the chain with an allowed url, the redirects after that could be
+   *   considered and implementation detail of the originating url's content
+   *   owner.  The counter argument is that if we only allow some URLs, that we
+   *   should ensure that nothing in the redirect chain is from a site other than
+   *   those.  This can be checked for after this call by using the addresses
+   *   property of the resolution object.
+   */
+  async validateInputURL (firstUrl) {
```

Comment:
1) this is generic logic, not backends specific. So feels like a very wrong a confusing place to put it.
Let's move it.

2) Do we actually need redirect logic? The URLs you get from queue will always point to the canonical S3 bucket.
They won't do any redirects. 

3) If we want redirect logic for completeness, isn't a more correct logic to simply cache the redirect answer.

----
I strongly advice that we don't do redirect logic... if we follow the redirect and store the result, then we risk duplicating results.
If we cache the redirect url, in the spirit of being a generic caching solution then we really ought to honor cache policies too. We could do this by storing such information in meta-data tags.

At the end of the day, I suggest we don't aim for this to be a generic caching proxy solution. But rather a cloud-mirror for immutable object-stores like S3. Ie. we're mirroring S3 into other object-stores on-demand.
I'm pretty sure that's also what you've been aiming for here, so let's not follow redirects, and just return 400 if the source url causes a redirect...